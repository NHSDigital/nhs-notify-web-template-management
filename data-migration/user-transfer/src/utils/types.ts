export type Parameters = {
  environment: string;
  component: string;
  accessKeyId: string;
  secretAccessKey: string;
  userPoolId: string;
  region: string;
  sessionToken: string;
  flag?: string;
};

export type UserData = {
  username: string;
  /*
   * extracted from the user's group starting with 'client:'
   */
  clientId: string;
  /*
   * extracted from the user's 'sub' attribute
   */
  userId: string;
};

export type Template = {
  id: string;
  owner: string;
};

type ToFrom = {
  from: string;
  to: string;
};

export type TransferStage =
  | 's3:copy'
  | 'ddb:transfer'
  | 's3:delete'
  | 'initial'
  | 'finished';

export type TransferStatus = 'success' | 'failed' | 'migrate';

export type UserTransferPlanItem = {
  templateId: string;
  status: TransferStatus;
  stage: TransferStage;
  reason?: string;
  ddb: {
    owner: ToFrom;
  };
  s3: {
    files: ToFrom[];
  };
};

export type UserTransferPlan = {
  total: number;
  tableName: string;
  bucketName: string;
  migrate: {
    count: number;
    plans: UserTransferPlanItem[];
  };
  orphaned: {
    count: number;
    templateIds: string[];
  };
};

export type UserTransferPlanItemResult = {
  success: boolean;
  stage: TransferStage;
  reasons?: string[];
};
