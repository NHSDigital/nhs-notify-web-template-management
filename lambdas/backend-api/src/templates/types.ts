export type User = {
  userId: string;
  clientId: string | undefined;
};
