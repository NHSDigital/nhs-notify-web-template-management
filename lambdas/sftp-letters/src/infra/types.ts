export type SftpSupplierConfig = {
  host: string;
  username: string;
  privateKey: string;
  hostKey: string;
  baseUploadDir: string;
  baseDownloadDir: string;
};
