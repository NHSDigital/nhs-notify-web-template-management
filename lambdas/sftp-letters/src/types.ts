import { Language, LetterType } from 'nhs-notify-backend-client';

export type ExpandedIdComponents = {
  clientId: string;
  templateId: string;
  campaignId: string;
  language: Language;
  letterType: LetterType;
};
