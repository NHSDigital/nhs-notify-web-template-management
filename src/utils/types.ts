export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
