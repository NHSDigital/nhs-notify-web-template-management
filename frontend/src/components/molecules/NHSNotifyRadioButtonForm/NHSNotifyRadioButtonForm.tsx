import { Radios, Fieldset, Button } from 'nhsuk-react-components';
import { FormState } from 'nhs-notify-web-template-management-utils';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';

export type NHSNotifyRadioButtonFormProps = {
  formId: string;
  radiosId: string;
  action: string | ((payload: FormData) => void);
  state: FormState;
  pageHeading: string;
  options: {
    id: string;
    text: string;
    checked?: boolean;
  }[];
  buttonText: string;
  hint?: string;
  legend?: {
    isPgeHeading: boolean;
    size: 'l' | 'm' | 's';
  };
  learnMoreLink?: string;
  learnMoreText?: string;
};

export const NHSNotifyRadioButtonForm = ({
  formId,
  radiosId,
  action,
  state,
  pageHeading,
  options,
  buttonText,
  legend = { isPgeHeading: true, size: 'l' },
  hint = '',
  learnMoreLink = '',
  learnMoreText = '',
}: NHSNotifyRadioButtonFormProps) => (
  <NHSNotifyFormWrapper action={action} formId={formId}>
    <Fieldset>
      <Fieldset.Legend
        data-testid={`${radiosId}-form__legend`}
        isPageHeading={legend.isPgeHeading}
        size={legend.size}
      >
        {pageHeading}
      </Fieldset.Legend>
      {learnMoreLink && learnMoreText && (
        <p>
          <a href={learnMoreLink} target='_blank' rel='noopener noreferrer'>
            {learnMoreText}
          </a>
        </p>
      )}
      <Radios
        id={radiosId}
        hint={hint}
        error={state.validationError?.fieldErrors[radiosId]?.join(', ')}
        errorProps={{ id: `${radiosId}-error-message` }}
      >
        {options.map(({ id, text, checked }) => (
          <Radios.Radio
            value={id}
            id={`${radiosId}-${id}`}
            data-testid={`${id}-radio`}
            key={`${id}-radio`}
            defaultChecked={checked}
          >
            {text}
          </Radios.Radio>
        ))}
      </Radios>
    </Fieldset>
    <Button
      type='submit'
      data-testid='submit-button'
      id={`${formId}-submit-button`}
    >
      {buttonText}
    </Button>
  </NHSNotifyFormWrapper>
);
