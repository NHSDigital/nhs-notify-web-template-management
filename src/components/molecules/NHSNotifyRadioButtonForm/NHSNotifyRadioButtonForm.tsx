import { Radios, Fieldset, Button } from 'nhsuk-react-components';
import { FormState } from '../../../utils/types';
import { NHSNotifyFormWrapper } from '../NHSNotifyFormWrapper/NHSNotifyFormWrapper';

export type NHSNotifyRadioButtonFormProps = {
  formId: string;
  radiosId: string;
  action: string | ((payload: FormData) => void);
  state: FormState;
  pageHeading: string;
  options: {
    id: string;
    text: string;
  }[];
  buttonText: string;
  hint?: string;
};

export const NHSNotifyRadioButtonForm = ({
  formId,
  radiosId,
  action,
  state,
  pageHeading,
  options,
  buttonText,
  hint,
}: NHSNotifyRadioButtonFormProps) => (
  <NHSNotifyFormWrapper action={action} formId={formId}>
    <Fieldset>
      <Fieldset.Legend isPageHeading size='l'>
        {pageHeading}
      </Fieldset.Legend>
      <Radios
        id={radiosId}
        hint={hint}
        error={state.validationError?.fieldErrors[radiosId]?.join(', ')}
        errorProps={{ id: `${radiosId}-error-message` }}
      >
        {options.map(({ id, text }) => (
          <Radios.Radio
            value={id}
            id={`${radiosId}-${id}`}
            data-testid={`${id}-radio`}
            key={`${id}-radio`}
          >
            {text}
          </Radios.Radio>
        ))}
      </Radios>
    </Fieldset>
    <Button
      type='submit'
      data-testid='submit-button'
      id={`${formId}-submit-button`}
    >
      {buttonText}
    </Button>
  </NHSNotifyFormWrapper>
);
