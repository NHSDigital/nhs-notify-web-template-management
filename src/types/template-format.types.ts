export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
