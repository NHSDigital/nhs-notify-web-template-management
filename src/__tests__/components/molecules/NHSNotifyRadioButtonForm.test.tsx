import { mockDeep } from 'jest-mock-extended';
import { render } from '@testing-library/react';
import { NHSNotifyRadioButtonForm } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';
import { FormState } from '@utils/types';

test('Renders NHSNotifyRadioButtonForm correctly without errors', () => {
  const container = render(
    <NHSNotifyRadioButtonForm
      formId='form-id'
      radiosId='radios-id'
      action='/action'
      state={mockDeep<FormState>({
        validationError: {
          fieldErrors: {},
          formErrors: [],
        },
      })}
      pageHeading='Page heading'
      options={[
        { id: 'option-1', text: 'option 1' },
        { id: 'option-2', text: 'option 2' },
      ]}
      buttonText='Continue'
      hint='Example hint'
    />
  );

  expect(container.asFragment()).toMatchSnapshot();
});

test('Renders NHSNotifyRadioButtonForm correctly with errors', () => {
  const container = render(
    <NHSNotifyRadioButtonForm
      formId='form-id'
      radiosId='radios-id'
      action='/action'
      state={mockDeep<FormState>({
        validationError: {
          fieldErrors: {
            'radios-id': ['Field error'],
          },
          formErrors: ['Form error'],
        },
      })}
      pageHeading='Page heading'
      options={[
        { id: 'option-1', text: 'option 1' },
        { id: 'option-2', text: 'option 2' },
      ]}
      buttonText='Continue'
    />
  );

  expect(container.asFragment()).toMatchSnapshot();
});

test('Renders NHSNotifyRadioButtonForm - handles validation errors for other fields', () => {
  const container = render(
    <NHSNotifyRadioButtonForm
      formId='form-id'
      radiosId='radios-id'
      action='/action'
      state={mockDeep<FormState>({
        validationError: {
          fieldErrors: {
            'radios-id-2': ['Field error'],
          },
          formErrors: ['Form error'],
        },
      })}
      pageHeading='Page heading'
      options={[
        { id: 'option-1', text: 'option 1' },
        { id: 'option-2', text: 'option 2' },
      ]}
      buttonText='Continue'
    />
  );

  expect(container.asFragment()).toMatchSnapshot();
});

test('Renders NHSNotifyRadioButtonForm - renders without validation error field', () => {
  const container = render(
    <NHSNotifyRadioButtonForm
      formId='form-id'
      radiosId='radios-id'
      action='/action'
      state={mockDeep<FormState>({})}
      pageHeading='Page heading'
      options={[
        { id: 'option-1', text: 'option 1' },
        { id: 'option-2', text: 'option 2' },
      ]}
      buttonText='Continue'
    />
  );

  expect(container.asFragment()).toMatchSnapshot();
});
