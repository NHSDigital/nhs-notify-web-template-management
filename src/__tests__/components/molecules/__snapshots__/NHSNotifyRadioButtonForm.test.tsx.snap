// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders NHSNotifyRadioButtonForm - handles validation errors for other fields 1`] = `
<DocumentFragment>
  <form
    action="/action"
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="form-id"
    />
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
          data-testid="radios-id-form__legend"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Page heading
          </h1>
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-radios"
            id="radios-id"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="radios-id-option-1"
                name="radios-id"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-1"
                id="radios-id-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="radios-id-option-2"
                name="radios-id"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-2"
                id="radios-id-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="form-id-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;

exports[`Renders NHSNotifyRadioButtonForm - renders without validation error field 1`] = `
<DocumentFragment>
  <form
    action="/action"
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="form-id"
    />
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
          data-testid="radios-id-form__legend"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Page heading
          </h1>
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-radios"
            id="radios-id"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="radios-id-option-1"
                name="radios-id"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-1"
                id="radios-id-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="radios-id-option-2"
                name="radios-id"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-2"
                id="radios-id-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="form-id-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;

exports[`Renders NHSNotifyRadioButtonForm correctly with errors 1`] = `
<DocumentFragment>
  <form
    action="/action"
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="form-id"
    />
    <div
      class="nhsuk-form-group nhsuk-form-group--error"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
          data-testid="radios-id-form__legend"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Page heading
          </h1>
        </legend>
        <div
          class="nhsuk-form-group nhsuk-form-group--error"
        >
          <span
            class="nhsuk-error-message"
            id="radios-id-error-message"
          >
            <span
              class="nhsuk-u-visually-hidden"
            >
              Error: 
            </span>
            Field error
          </span>
          <div
            aria-describedby="radios-id--error-message"
            class="nhsuk-radios"
            id="radios-id"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="radios-id-option-1"
                name="radios-id"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-1"
                id="radios-id-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="radios-id-option-2"
                name="radios-id"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-2"
                id="radios-id-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="form-id-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;

exports[`Renders NHSNotifyRadioButtonForm correctly without errors 1`] = `
<DocumentFragment>
  <form
    action="/action"
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="form-id"
    />
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--l"
          data-testid="radios-id-form__legend"
        >
          <h1
            class="nhsuk-fieldset__heading"
          >
            Page heading
          </h1>
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-hint"
            id="radios-id--hint"
          >
            Example hint
          </div>
          <div
            aria-describedby="radios-id--hint"
            class="nhsuk-radios"
            id="radios-id"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="radios-id-option-1"
                name="radios-id"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-1"
                id="radios-id-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="radios-id-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="radios-id-option-2"
                name="radios-id"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="radios-id-option-2"
                id="radios-id-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="form-id-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;
